### PR TITLE
[SPARK-39217][SQL] Makes DPP support the pruning side has Union

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Makes DPP support the pruning side has `Union`. For example:
```sql
SELECT f.store_id,
       f.date_id,
       s.state_province
FROM (SELECT 4 AS store_id,
               date_id,
               product_id
      FROM   fact_sk
      WHERE  date_id >= 1300
      UNION ALL
      SELECT   store_id,
               date_id,
               product_id
      FROM   fact_stats
      WHERE  date_id <= 1000) f
JOIN dim_store s
ON f.store_id = s.store_id
WHERE s.country IN ('US', 'NL')
```
After this PR:
image

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

**Comment by wangyum** [comment]
A case from production:
image

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/CleanupDynamicPruningFilters.scala [hunk 1]
```diff
@@ -54,7 +54,8 @@ object CleanupDynamicPruningFilters extends Rule[LogicalPlan] with PredicateHelp
         val newCondition = condition.transformWithPruning(
           _.containsPattern(DYNAMIC_PRUNING_SUBQUERY)) {
           case dynamicPruning: DynamicPruningSubquery
-              if unnecessaryPruningKeys.contains(dynamicPruning.pruningKey) =>
+              if unnecessaryPruningKeys.contains(dynamicPruning.pruningKey) ||
+                dynamicPruning.pruningKey.references.isEmpty =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/CleanupDynamicPruningFilters.scala:58]: how is this change related to Union?

**Comment by cloud-fan** [comment]
sorry for the late review, the change looks reasonable to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/CleanupDynamicPruningFilters.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/CleanupDynamicPruningFilters.scala:58]: This issue has fixed: https://github.com/apache/spark/pull/36724

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -140,6 +140,15 @@ trait PredicateHelper extends AliasHelper with Logging {
         findExpressionAndTrackLineageDown(replaceAlias(exp, aliasMap), a.child)
       case l: LeafNode if exp.references.subsetOf(l.outputSet) =>
         Some((exp, l))
+      case u: Union =>
+        val index = u.output.indexWhere(_.semanticEquals(exp))
+        if (index > -1) {
+          u.children
+            .flatMap(child => findExpressionAndTrackLineageDown(child.output(index), child))
+            .sortBy(_._2.stats.sizeInBytes).lastOption
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:148]: can you explain a bit more about this sorting part?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:148]: In general, tables with larger sizes are more likely to be partition tables.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:148]: and we only add DPP filter in one child of Union?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:148]: For all child of Union expect the references is empty. For example: https://github.com/apache/spark/pull/36724

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:148]: shall we do the same for `case other =>`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:148]: After reading the code more, I don't think we have to use a heuristic here. We can let caller side to pass a callback function so that we can go to all the branches to find partitioned tables.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: OK

**Comment by wangyum** [comment]
@cloud-fan Do you have more comments?
